### ISSUE TITLE
Empty list in dataclass initializers is misinterpreted

### ISSUE THREAD

**Issue by rebcabin** [lcompilers/lpython#1990]
LPython gives what I think is an incorrect error message.

repro:

```python
from lpython import dataclass

@dataclass
class Pattern:
    _foo : list[tuple[str]]

foo : list[tuple[str]] = []

pat : Pattern = Pattern([])
```

CPython: 

```bash
└─(13:05:49 on brian-lasr ✹ ✭)──> PYTHONPATH='../../../src/runtime/lpython' python ./test_empty_list.py                                                   ──(Wed,Jun21)─┘
(lp) ┌─(~/CLionProjects/lpython/lasr/LP-pycharm/Issue1989)───────────────────────────────────────────────────────────────────────────────────────────────(brian@Golf37:s014)─┐
└─(13:06:25 on brian-lasr ✹ ✭)──>  
```

LPython:

```bash
(lp) ┌─(~/CLionProjects/lpython/lasr/LP-pycharm/Issue1989)───────────────────────────────────────────────────────────────────────────────────────────────(brian@Golf37:s014)─┐
└─(13:06:25 on brian-lasr ✹ ✭)──> ../../../src/bin/lpython -I. ./test_empty_list.py                                                                       ──(Wed,Jun21)─┘
semantic error: Type mismatch in procedure call; the types must be compatible
 --> ./test_empty_list.py:9:25
  |
9 | pat : Pattern = Pattern([])
  |                         ^^ type mismatch (passed argument type is list[Pattern] but required type is list[tuple[str]])
  |
5 |     _foo : list[tuple[str]]
  |     ^^^^^^^^^^^^^^^^^^^^^^^ type mismatch (passed argument type is list[Pattern] but required type is list[tuple[str]])
```

**Comment by rebcabin** [comment]
putting some junk data in fixes it for LPython:

```python
from lpython import dataclass

@dataclass
class Pattern:
    _foo : list[tuple[str]]

foo : list[tuple[str]] = []

pat : Pattern = Pattern([('',)])
```

**Comment by rebcabin** [comment]
Notice that LPython has no problem with the assignment of an empty list to `foo`, which is outside a dataclass. LPython recognizes that the empty list `[]` has type `list[tuple[str]]` in that case.

**Comment by certik** [comment]
Thanks. I think the type of `[]` in `Pattern([])` should be `list[tuple[str]]`, but Python thinks it's `list[Pattern]`, which is wrong. I think that's the main bug here.
